### PR TITLE
coverity 1022151: Resource leak

### PR DESCRIPTION
**Pull request by bryancall**

### CODE REVIEW

File: lib/ts/ink_memory.h [hunk 1]
```diff
@@ -534,4 +535,12 @@ path_join(ats_scoped_str const &lhs, ats_scoped_str const &rhs)
 }
 #endif /* __cplusplus */
 
+// This is in C++14 and should removed when we support it
+template <typename T, typename... Args>
+std::unique_ptr<T>
+make_unique(Args &&... args)
+{
+  return std::unique_ptr<T>(new T(std::forward<Args>(args)...));
+}
```
Comment (jpeach) [inline lib/ts/ink_memory.h]: Put this in a
```C
#if __cplusplus < 201402L
...
#endif
```

File: proxy/RegressionSM.cc [hunk 2]
```diff
@@ -273,10 +275,16 @@ struct ReRegressionSM : public RegressionSM {
...
+
   RegressionSM *top_sm = r_sequential(
-    t, r_parallel(t, new ReRegressionSM(t), new ReRegressionSM(t), nullptr),
-    r_sequential(t, new ReRegressionSM(t), new ReRegressionSM(t), nullptr), r_parallel(t, 3, new ReRegressionSM(t)),
-    r_sequential(t, 3, new ReRegressionSM(t)),
-    r_parallel(t, r_sequential(t, 2, new ReRegressionSM(t)), r_parallel(t, 2, new ReRegressionSM(t)), nullptr), nullptr);
+    t, r_parallel(t, state_machines[0].get(), state_machines[1].get(), nullptr),
+    r_sequential(t, state_machines[2].get(), state_machines[3].get(), nullptr), r_parallel(t, 3, state_machines[4].get()),
+    r_sequential(t, 3, state_machines[5].get()),
+    r_parallel(t, r_sequential(t, 2, state_machines[6].get()), r_parallel(t, 2, state_machines[7].get()), nullptr), nullptr);
```
Comment (jpeach) [inline proxy/RegressionSM.cc]: Wow `clang-format` really munged that. Can we format this so there is one call per line,
```C
RegressionSM *top_sm = r_sequential(t,
  r_parallel(t, state_machines[0].get(), state_machines[1].get(), nullptr),
  r_sequential(t, state_machines[2].get(), state_machines[3].get(), nullptr),
  r_parallel(t, 3, state_machines[4].get()),
  ..
);
```

File: proxy/RegressionSM.cc [hunk 3]
```diff
@@ -273,10 +275,16 @@ struct ReRegressionSM : public RegressionSM {
 
 REGRESSION_TEST(RegressionSM)(RegressionTest *t, int /* atype ATS_UNUSED */, int *pstatus)
 {
+  // create 8 state machines
+  std::vector<std::unique_ptr<ReRegressionSM>> state_machines;
+  for (auto i = 0; i < 8; i++) {
+    state_machines.push_back(make_unique<ReRegressionSM>(t));
+  }
```
Comment (jpeach) [inline proxy/RegressionSM.cc]: Is this the right fix? Isn't the root cause that `RegressionSM` doesn't delete it's children? At least we should have a comment here explaining what is going on and why this doesn't end up deleting state machines before they are done.

**Comment by bryancall** [comment]
@jpeach I looked into this deeper and found that it does delete its children, but is unclear from the interface.  I am working on a change to use vectors in RegressionSM instead of DynArray.

**Pull request closed without merging**